src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1413,86 +1389,68 @@ final ForkJoinTask<?> peek() {
 
         /**
          * Polls for a task. Used only by non-owners.
-         *
-         * @param pool if nonnull, pool to signal if more tasks exist
          */
-        final ForkJoinTask<?> poll(ForkJoinPool pool) {
-            for (;;) {
-                ForkJoinTask<?>[] a = array;
-                int b = base, cap, k;
-                if (a == null || (cap = a.length) <= 0)
-                    break;
-                ForkJoinTask<?> t = a[k = b & (cap - 1)];
-                U.loadFence();
-                if (base == b) {
-                    Object o;
-                    int nb = b + 1, nk = nb & (cap - 1);
-                    if (t == null)
-                        o = a[k];
-                    else if (t == (o = U.compareAndExchangeReference(
-                                       a, slotOffset(k), t, null))) {
+        final ForkJoinTask<?> poll() {
+            if ((phase & (IDLE | 1)) == 0 || top - base > 0) {
+                for (boolean stalled = false;;) {
+                    int cap, b, k, nb; ForkJoinTask<?>[] a;
+                    if ((a = array) == null || (cap = a.length) <= 0)
+                        break;
+                    long kp = slotOffset(k = (cap - 1) & (b = base));
+                    int nk = (nb = b + 1) & (cap - 1); // next slot
+                    int sk = (b + 2) & (cap - 1);      // 2nd slot ahead
+                    ForkJoinTask<?> t = a[k];
+                    U.loadFence();
+                    if (base != b)
+                        ;                              // inconsistent
```

Comment:
@DougLea This seems like it could be a `continue;`, which might be easier to reason about?

Reply:
You'd think these would be equivalent, but not always so, because of impact on safepoints. More compiler-whispering. But it can be a little more readable by converting to nested if with same effect. Will do.